pkg/cmd/roachtest-selector/sfselector/selector.go

```diff
@@ -0,0 +1,298 @@
+// 2. uploading the results in CSV format to cloud storage
+func SelectTests(ctx context.Context, req *SelectTestsReq) error {
+	fmt.Printf("getting test details.\n")
+	r, err := getTestDetails(ctx, req)
+	if err != nil {
+		return err
+	}
+	csvBuffer, err := getCSVInMemory(r)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("uploading %s cloud test details to the cloud...\n", req.Cloud)
+	if req.DryRun {
+		fmt.Println(csvBuffer.String())
+		return nil
+	}
+	return uploadCSVToGCS(csvBuffer, req.Cloud, req.Suite)
+}
+
+// getTestDetails gets the test details from snowflake
+func getTestDetails(ctx context.Context, req *SelectTestsReq) (*testDetails, error) {
+	if err := req.validate(); err != nil {
+		return nil, err
+	}
+	db, err := getConnect(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = db.Close() }()
+	statement, err := db.Prepare(preparedQuery)
+	if err != nil {
+		return nil, err
+	}
+	// add the parameters in sequence
+	rows, err := statement.QueryContext(context.Background(),
+		req.ForPastDays*-1, fmt.Sprintf("%%%s - %s%%", suites[req.Suite], req.Cloud),
+		req.FirstRunOn*-1, req.LastRunOn*-1)
+	if err != nil {
+		return nil, err
+	}
+	// All the column headers
+	colHeaders, err := rows.Columns()
+	if err != nil {
+		return nil, err
+	}
+	// Will be used to read data while iterating rows.
+	colPointers := make([]interface{}, len(colHeaders))
+	colContainer := make([]string, len(colHeaders))
+	for i := range colPointers {
+		colPointers[i] = &colContainer[i]
+	}
+	// skipped tests are maintained separately
+	// this helps in considering them for running based on selectFromSuccessPct
+	skippedTests := make([][]string, 0)
+	td := &testDetails{
+		columnHeaders: colHeaders,
+		testInfos:     make([][]string, 0),
+	}
+	selectedTestCount := 0 // used for printing the number of tests selected
+	for rows.Next() {
+		_ = rows.Scan(colPointers...)
+		testInfos := make([]string, len(colContainer))
+		copy(testInfos, colContainer)
+		if testInfos[1] == "no" {
+			// skipped based on query
+			skippedTests = append(skippedTests, testInfos)
+		} else {
+			// selected for running
+			td.testInfos = append(td.testInfos, testInfos)
+			selectedTestCount++
+		}
+	}
+	if req.SelectFromSuccessPct > 0 && len(skippedTests) > 0 {
+		// need to select some tests from th success list (skipped tests)
+		numberOfTestsToSelect := len(skippedTests) * req.SelectFromSuccessPct / 100
+		// the tests are sorted by the number of runs. So, simply iterate over teh list
+		// and select the first count of "numberOfTestsToSelect"
+		for i := range skippedTests {
+			if i == numberOfTestsToSelect {
+				break
+			}
+			// mark the skipped tests as selected="yes"
+			skippedTests[i][1] = "yes"
+			selectedTestCount++
+		}
+	}
+	// add all the skipped tests (some may have been selected based on selectFromSuccessPct)
+	td.testInfos = append(td.testInfos, skippedTests...)
+	fmt.Printf("%d out of %d tests selected!\n", selectedTestCount, len(td.testInfos))
```

Comment:
Nit: this logging should be in the outer layers (i.e., `roachtest` in this case). This is a library call and it's nice to avoid as much side-effect as possible.